lib/wallet/src/Cardano/Wallet/Delegation.hs

```diff
@@ -208,3 +217,42 @@ guardVoting
 guardVoting optionalDelegationAction votingSameAgainM = do
     when (isNothing optionalDelegationAction && (fst <$> votingSameAgainM) == Just True ) $
         Left $ ErrAlreadyVoted $ snd (fromJust votingSameAgainM)
+
+guardOnlyVoting
+    :: Write.IsRecentEra era
+    => Write.RecentEra era
+    -> (Bool,DRep)
+    -> Either ErrCannotVote ()
+guardOnlyVoting era votingSameAgain = do
+    when (fst votingSameAgain) $
+        Left $ ErrAlreadyVoted $ snd votingSameAgain
+
+    case era of
+        Write.RecentEraBabbage ->
+            Left ErrWrongEra
+        Write.RecentEraConway ->
+            Right ()
+
+joinDRepVotingAction
+    :: Write.IsRecentEra era
+    => Write.RecentEra era
+    -> DRep
+    -> W.WalletDelegation
+    -> Bool
+    -> Either ErrCannotVote Tx.VotingAction
+joinDRepVotingAction era action dlg stakeKeyIsRegistered =
+    second (const vAction) $ guardOnlyVoting era votingRequest
+  where
+    isDRepSame (W.Voting drep) = drep == action
+    isDRepSame (W.DelegatingVoting _ drep) = drep == action
+    isDRepSame _ = False
+
+    isSameNext (W.WalletDelegationNext _ deleg) = isDRepSame deleg
+
+    sameWalletDelegation (W.WalletDelegation current coming) =
+        isDRepSame current || any isSameNext coming
+
+    (vAction, votingRequest) =
+        if stakeKeyIsRegistered
+        then (Tx.Vote action, (sameWalletDelegation dlg, action))
+        else (Tx.VoteRegisteringKey action, (sameWalletDelegation dlg,action))
```

Comment:
```suggestion
joinDRepVotingAction
    :: Write.IsRecentEra era
    => Write.RecentEra era
    -> DRep
    -> W.WalletDelegation
    -> Bool
    -> Either ErrCannotVote Tx.VotingAction
joinDRepVotingAction era targetDRep dlg stakeKeyIsRegistered = do
    guardEraIsConway era
    when (sameWalletDelegation dlg) $
        Left $ ErrAlreadyVoted $ targetDRep
    pure votingAction
  where
    isDRepSame (W.Voting drep) = drep == targetDRep
    isDRepSame (W.DelegatingVoting _ drep) = drep == targetDRep
    isDRepSame _ = False

    isSameNext (W.WalletDelegationNext _ deleg) = isDRepSame deleg

    sameWalletDelegation (W.WalletDelegation current coming) =
        isDRepSame current || any isSameNext coming

    guardEraIsConway Write.RecentEraBabbage = Left WrongEra
    guardEraIsConway Write.RecentEraConway = Right ()

    votingAction =
        if stakeKeyIsRegistered
        then Tx.Vote targetDRep
        else Tx.VoteRegisteringKey targetDRep
```

… and after separating the concerns into a single, we can now merge the definitions of the two functions and get rid of the `(Bool,DRep)` intermediate type while also making the logic more clear. 🤓

Reply:
good hint. Done!